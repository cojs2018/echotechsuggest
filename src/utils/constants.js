export const bookmarksAPI = 'https://5zcbdoz0h3.execute-api.eu-west-2.amazonaws.com/dev';
export const tagsAPI = 'https://141b51xkk2.execute-api.eu-west-2.amazonaws.com/dev';