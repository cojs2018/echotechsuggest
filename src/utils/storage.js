import { writeFile } from 'fs';

export async function createBookmark(bookmarkUrl) {
    try {
        let bookmarkUrlList = require('../../bookmarkUrls.json');

        let bookmarkLocation = bookmarkUrl;
        if(!bookmarkLocation.includes('http')) {
            bookmarkLocation = `http://${bookmarkLocation}`;
        }

        const newUrlData = new URL(bookmarkLocation);

        const bookmarkJSON = {
            path: bookmarkUrl,
            url: newUrlData.toJSON(),
            createdAt: Date.now()
        }
        const articleName = `${newUrlData.pathname} - ${newUrlData.hostname}`;

        bookmarkUrlList[articleName] = bookmarkJSON;

        writeFile(
            '../../bookmarkUrls.json',
            JSON.stringify(bookmarkJSON, ' ', 4)
        )

        return {
            status: 200,
            message: 'Bookmark created successfully!'
        }
    }
    catch (error) {
        throw new Error(JSON.stringify({ 
            status: 500,
            message: 'Internal service error: Could not create bookmark!' 
        }));
    }
}

export async function listBookmarks() {
    const bookmarkUrlList = require('../../bookmarkUrls.json');
    return bookmarkUrlList;
}

export async function deleteBookmark(bookmarkName) {
    try {
        let bookmarkUrlList = require('../../bookmarkUrls.json');
        
        if(!bookmarkUrlList[bookmarkName]) {
            return {
                status: 200,
            }
        }

        delete bookmarkUrlList[bookmarkName]

        writeFile(
            '../../bookmarkUrls.json',
            JSON.stringify(bookmarkJSON, ' ', 4)
        )

        return {
            status: 200,
            message: 'Bookmark deleted successfully!'
        }
    }
    catch (error) {
        throw new Error(JSON.stringify({ 
            status: 500,
            message: 'Internal service error: Could not delete bookmark!' 
        }));
    }
}