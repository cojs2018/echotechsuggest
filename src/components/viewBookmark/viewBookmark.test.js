jest.mock('../../utils/storage');
jest.mock('./tagList/tagList');

import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import ViewBookmark from './viewBookmark';
import TagList from './tagList/tagList';
import { getBookmark, deleteBookmark } from '../../utils/storage';
import { View } from 'react-native';
import { Text } from 'react-native-paper';

const tags = [
    {
        "tagId": "94972406-7158-4a27-8a44-52a33b854a74",
        "value": "Artificial Precipitation",
        "featured": 1
    },
    {
        "tagId": "18a2d966-fe32-4d05-9543-fcee703f718c",
        "value": "Energy",
        "featured": 1
    },
    {
        "tagId": "52d91ec3-a18b-4864-8438-56cfd759528c",
        "value": "Quantum Computing",
        "featured": 2
    },
    {
        "tagId": "8b48abe0-98df-4c33-99d4-543ff3f4bf54",
        "value": "Biotechnology",
        "featured": 1
    }
]

getBookmark.mockReturnValue(new Promise((resolve, reject) => {
    resolve({
        bookmarkId: '047ec796-2eca-417e-865a-e2e187b97689',
        articleName: 'breakthrough-for-healthcare-agriculture-energy-artificial-intelligence-reveals-recipe-for-building-artificial-proteins - scitechdaily',
        path: 'https://scitechdaily.com/breakthrough-for-healthcare-agriculture-energy-artificial-intelligence-reveals-recipe-for-building-artificial-proteins/',
        url: {},
        title: 'Breakthrough for Healthcare, Agriculture, Energy: Artificial Intelligence Reveals Recipe for Building Artificial Proteins',
        description: 'Proteins are essential to cells, carrying out complex tasks and catalyzing chemical reactions. Scientists and engineers have long sought to harness this power by designing artificial proteins that can perform new tasks, like treat disease, capture carbon or harvest energy, but many of the processes',
        tags,
        images: [],
    })
}));

TagList.mockReturnValue(<View>
    {tags.map(tag => <Text>{tag.value}</Text>)}
</View>);

describe('ViewBookmark', () => {
    const props = {
        bookmarkIdSelected: '047ec796-2eca-417e-865a-e2e187b97689',
        setPage: jest.fn(async () => {}),
    }

    it('reders correctly', () => {
        const tree = renderer
            .create(<ViewBookmark {...props} />)
            .toJSON();

        expect(tree).toMatchSnapshot();
    })
});

test('Close', async () => {
    let page = 3;

    const props = {
        bookmarkIdSelected: '047ec796-2eca-417e-865a-e2e187b97689',
        setPage: jest.fn(async (newPage) => page = newPage),
    }

    const wrapper = mount(<ViewBookmark {...props} />);

    const closeIcon = wrapper
        .find('IconButton')
        .filterWhere(iconButton => iconButton.props().icon === 'close')
        .first();

    closeIcon.props().onPress();
    wrapper.update();

    await new Promise(resolve => setImmediate(resolve));
    expect(page).toStrictEqual(2);
})

test('Delete', async () => {
    let page = 3;

    const props = {
        bookmarkIdSelected: '047ec796-2eca-417e-865a-e2e187b97689',
        setPage: jest.fn(async (newPage) => page = newPage),
    }

    /* let isLoading = false;
    let details = {
        bookmarkId: '047ec796-2eca-417e-865a-e2e187b97689',
        articleName: '',
        path: 'No Path',
        url: 'No URL',
        title: 'No Title',
        description: 'No description',
        tags: [],
        images: [],
    };
    let alertMessage = {
        status: 'none',
        message: ''
    };
    let visible = false;

    const setIsLoading = jest.fn(async (bool) => isLoading = bool);
    const setDetails = jest.fn(async (loadedDetails) => details = loadedDetails);
    const setAlertMessage = jest.fn(async (newMessage) => alertMessage = newMessage);
    const setVisible = jest.fn(async (bool) => visible = bool);

    React.useEffect = jest.fn()
        .mockImplementationOnce((effect, deps = null) => effect());
    
    React.useState = jest.fn()
        .mockReturnValueOnce([isLoading, setIsLoading])
        .mockReturnValueOnce([details, setDetails])
        .mockReturnValueOnce([alertMessage, setAlertMessage])
        .mockReturnValueOnce([visible, setVisible]); */

    deleteBookmark.mockReturnValueOnce(new Promise((resolve, reject) => {
        resolve({
            status: 200,
            message: `Bookmark ${props.bookmarkIdSelected} Deleted Successfully!`
        });
    }));

    const wrapper = mount(<ViewBookmark {...props} />);

    const deleteIcon = wrapper
        .find('IconButton')
        .filterWhere(iconButton => iconButton.props().icon === 'delete')
        .first();

    deleteIcon.props().onPress();
    wrapper.update();

    await new Promise(resolve => setImmediate(resolve));
    //expect(alertMessage.status).toStrictEqual('success');
})