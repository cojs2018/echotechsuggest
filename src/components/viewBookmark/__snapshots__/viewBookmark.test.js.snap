// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewBookmark reders correctly 1`] = `
<View>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#6200ee",
        "elevation": 4,
        "flexDirection": "row",
        "height": 56,
        "paddingHorizontal": 4,
        "shadowColor": "#000000",
        "shadowOffset": Object {
          "height": 3,
          "width": 0,
        },
        "shadowOpacity": 0.24,
        "shadowRadius": 4,
      }
    }
  >
    <View
      accessible={true}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "flex": 1,
            "paddingHorizontal": 12,
          },
          Array [
            false,
            false,
            undefined,
          ],
        ]
      }
    >
      <Text
        accessibilityRole="header"
        accessibilityTraits="header"
        accessible={true}
        numberOfLines={1}
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "System",
              "fontWeight": "400",
              "textAlign": "left",
            },
            Array [
              Object {
                "color": "#ffffff",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              Object {
                "fontSize": 17,
              },
              undefined,
            ],
          ]
        }
      >
        047ec796-2eca-417e-865a-e2e187b97689
      </Text>
    </View>
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "disabled": undefined,
        }
      }
      accessible={true}
      focusable={true}
      hitSlop={
        Object {
          "bottom": 6,
          "left": 6,
          "right": 6,
          "top": 6,
        }
      }
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Array [
            Object {
              "alignItems": "center",
              "justifyContent": "center",
              "margin": 6,
              "overflow": "hidden",
            },
            Object {
              "borderRadius": 18,
              "height": 36,
              "width": 36,
            },
            undefined,
            undefined,
          ],
        ]
      }
    >
      <View>
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "justifyContent": "center",
              },
              Object {
                "height": 24,
                "width": 24,
              },
            ]
          }
        >
          <View
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "opacity": 1,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "transform": Array [
                  Object {
                    "rotate": "0deg",
                  },
                ],
              }
            }
          >
            <Text
              accessibilityElementsHidden={true}
              allowFontScaling={false}
              importantForAccessibility="no-hide-descendants"
              pointerEvents="none"
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "#ffffff",
                    "fontSize": 24,
                  },
                  Array [
                    Object {
                      "lineHeight": 24,
                      "transform": Array [
                        Object {
                          "scaleX": 1,
                        },
                      ],
                    },
                    Object {
                      "backgroundColor": "transparent",
                    },
                  ],
                  Object {
                    "fontFamily": "Material Design Icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              󰆴
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "disabled": undefined,
        }
      }
      accessible={true}
      focusable={true}
      hitSlop={
        Object {
          "bottom": 6,
          "left": 6,
          "right": 6,
          "top": 6,
        }
      }
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Array [
            Object {
              "alignItems": "center",
              "justifyContent": "center",
              "margin": 6,
              "overflow": "hidden",
            },
            Object {
              "borderRadius": 18,
              "height": 36,
              "width": 36,
            },
            undefined,
            undefined,
          ],
        ]
      }
    >
      <View>
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "justifyContent": "center",
              },
              Object {
                "height": 24,
                "width": 24,
              },
            ]
          }
        >
          <View
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "opacity": 1,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "transform": Array [
                  Object {
                    "rotate": "0deg",
                  },
                ],
              }
            }
          >
            <Text
              accessibilityElementsHidden={true}
              allowFontScaling={false}
              importantForAccessibility="no-hide-descendants"
              pointerEvents="none"
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "#ffffff",
                    "fontSize": 24,
                  },
                  Array [
                    Object {
                      "lineHeight": 24,
                      "transform": Array [
                        Object {
                          "scaleX": 1,
                        },
                      ],
                    },
                    Object {
                      "backgroundColor": "transparent",
                    },
                  ],
                  Object {
                    "fontFamily": "Material Design Icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              󰅖
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "flex-start",
        "flex": 1,
        "flexDirection": "column",
        "height": "100%",
        "justifyContent": "flex-start",
        "width": "97%",
      }
    }
  >
    <View />
    <View>
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "System",
              "fontWeight": "400",
              "textAlign": "left",
            },
            Object {
              "fontWeight": "bold",
              "paddingBottom": 10,
              "paddingTop": 20,
            },
          ]
        }
      >
        Title
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "System",
              "fontWeight": "400",
              "textAlign": "left",
            },
            undefined,
          ]
        }
      >
        No Title
      </Text>
    </View>
    <View>
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "System",
              "fontWeight": "400",
              "textAlign": "left",
            },
            Object {
              "fontWeight": "bold",
              "paddingBottom": 10,
              "paddingTop": 20,
            },
          ]
        }
      >
        Description
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "System",
              "fontWeight": "400",
              "textAlign": "left",
            },
            undefined,
          ]
        }
      >
        No description
      </Text>
    </View>
    <View>
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "System",
              "fontWeight": "400",
              "textAlign": "left",
            },
            Object {
              "fontWeight": "bold",
              "paddingBottom": 10,
              "paddingTop": 20,
            },
          ]
        }
      >
        Path
      </Text>
      <Text
        accessibilityRole="link"
        dataDetectorType="link"
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "System",
              "fontWeight": "400",
              "textAlign": "left",
            },
            undefined,
          ]
        }
      >
        No Path
      </Text>
    </View>
    <View>
      <View>
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
                "textAlign": "left",
              },
              undefined,
            ]
          }
        >
          Artificial Precipitation
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
                "textAlign": "left",
              },
              undefined,
            ]
          }
        >
          Energy
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
                "textAlign": "left",
              },
              undefined,
            ]
          }
        >
          Quantum Computing
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
                "textAlign": "left",
              },
              undefined,
            ]
          }
        >
          Biotechnology
        </Text>
      </View>
    </View>
  </View>
</View>
`;
